src/modem/mod.rs

```diff
@@ -308,6 +349,54 @@ impl<'c, P: ModemPower> Modem<'c, P> {
         Ok(())
     }
 
+    /// Resets the network priority to the priority provided when initializing [Modem::init] with [NetworkModeConfig::Automatic]
+    ///
+    /// If not initialized with [NetworkModeConfig::Automatic], this function has no effect
+    pub fn reset_network_priority(&mut self) {
+        self.current_network_priority = self.user_network_priority.clone();
+    }
+
+    /// Connect to the first available radio access technology (RAT).
+    /// If connected using LTE-CatM or GSM, set that RAT as first priority for next registration attempt
+    async fn automatic_registration(
+        &self,
+        commands: &CommandRunnerGuard<'_>,
+    ) -> Result<SystemMode, Error> {
+        for mode in &self.current_network_priority {
+            match mode {
+                SystemMode::LteCatM1 => {
+                    commands.run(cnmp::SetNetworkMode(NetworkMode::Lte)).await?;
+                    commands.run(cmnb::SetNbMode(NbMode::CatM)).await?;
+                }
+                SystemMode::Gsm => {
+                    commands.run(cnmp::SetNetworkMode(NetworkMode::Gsm)).await?;
+                }
+                SystemMode::LteNbIot => {
+                    commands.run(cnmp::SetNetworkMode(NetworkMode::Lte)).await?;
+                    commands.run(cmnb::SetNbMode(NbMode::NbIot)).await?;
+                }
+                _ => {
+                    log::error!("Unsupported network mode: {:?}", mode);
+                    return Err(Error::Timeout);
+                }
+            }
+
+            log::info!("Trying {:?}...", mode);
+            match with_timeout(Duration::from_secs(2 * 60), self.wait_for_registration()).await {
+                Ok(Ok(_)) => {
+                    log::info!("Registered using {:?}", mode);
+                    return Ok(*mode);
+                }
+                Ok(Err(_)) => {
+                    // this should never happen since wait_for_registration timeout is longer than 2 min
+                }
+                Err(_) => {}
+            }
+        }
+
+        return Err(Error::Timeout);
```

Comment:
is it desirable to give up after trying for 6 minutes? an alternative might be to loop through the list forever and let the caller add a timeout if desired

Reply:
hm, idk. It becomes a "hidden" behaviour in that case, unless comments.. but do people read those? 🤔